File: src/streambench/sparkbench/src/main/scala/com/intel/hibench/streambench/spark/RunBench.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.hibench.streambench.spark

import com.intel.hibench.common.HiBenchConfig
import com.intel.hibench.streambench.common.metrics.{KafkaReporter, MetricsUtil, LatencyReporter}
import com.intel.hibench.streambench.common.{Platform, ConfigLoader, StreamBenchConfig}
import com.intel.hibench.streambench.spark.util.SparkBenchConfig
import com.intel.hibench.streambench.spark.application._
import kafka.serializer.StringDecoder
import org.apache.spark.SparkConf
import org.apache.spark.streaming.dstream.DStream
import org.apache.spark.streaming.kafka.KafkaUtils
import org.apache.spark.streaming.{Milliseconds, Seconds, StreamingContext}

/**
  * The entry point of Spark Streaming benchmark
  */
object RunBench {

  def main(args: Array[String]) {
    val conf = new ConfigLoader(args(0))

    // Load configuration
    val master = conf.getProperty(HiBenchConfig.SPARK_MASTER)
    val reportDir = conf.getProperty(HiBenchConfig.REPORT_DIR)

    val batchInterval = conf.getProperty(StreamBenchConfig.SPARK_BATCH_INTERVAL).toInt
    val receiverNumber = conf.getProperty(StreamBenchConfig.SPARK_RECEIVER_NUMBER).toInt
    val copies = conf.getProperty(StreamBenchConfig.SPARK_STORAGE_LEVEL).toInt
    val enableWAL = conf.getProperty(StreamBenchConfig.SPARK_ENABLE_WAL).toBoolean
    val checkPointPath = conf.getProperty(StreamBenchConfig.SPARK_CHECKPOINT_PATHL)
    val directMode = conf.getProperty(StreamBenchConfig.SPARK_USE_DIRECT_MODE).toBoolean
    val benchName = conf.getProperty(StreamBenchConfig.TESTCASE)
    val topic = conf.getProperty(StreamBenchConfig.KAFKA_TOPIC)
    val zkHost = conf.getProperty(StreamBenchConfig.ZK_HOST)
    val consumerGroup = conf.getProperty(StreamBenchConfig.CONSUMER_GROUP)
    val brokerList = if (directMode) conf.getProperty(StreamBenchConfig.KAFKA_BROKER_LIST) else ""
    val debugMode = conf.getProperty(StreamBenchConfig.DEBUG_MODE).toBoolean
    val recordPerInterval = conf.getProperty(StreamBenchConfig.DATAGEN_RECORDS_PRE_INTERVAL).toLong
    val intervalSpan: Int = conf.getProperty(StreamBenchConfig.DATAGEN_INTERVAL_SPAN).toInt

    // val separator = conf.getProperty(StreamBenchConfig.SEPARATOR)
    val recordCount = conf.getProperty("hibench.streamingbench.record_count").toLong
    val coreNumber = conf.getProperty("hibench.yarn.executor.num").toInt * conf.getProperty("hibench.yarn.executor.cores").toInt

    // val logPath = reportDir + "/streamingbench/spark/streambenchlog.txt"
    val reporterTopic = MetricsUtil.getTopic(Platform.SPARK, topic, recordPerInterval, intervalSpan)
    println("Reporter Topic" + reporterTopic)

    val probability = conf.getProperty(StreamBenchConfig.SAMPLE_PROBABILITY).toDouble
    // init SparkBenchConfig, it will be passed into every test case
    val config = SparkBenchConfig(master, benchName, batchInterval, receiverNumber, copies,
      enableWAL, checkPointPath, directMode, zkHost, consumerGroup, topic, reporterTopic,
      brokerList, recordCount, debugMode, coreNumber, probability)

    run(config)
  }

  private def run(config: SparkBenchConfig) {

    // select test case based on given benchName
    val testCase : BenchBase = config.benchName match {
      case "identity" => new Identity()
      case "repartition" => new Repartition()
      case "project" => new Project()
      case "wordcount" => new WordCount()
      case "statistics" =>
        // This test case will consume KMeansData instead of UserVisit as other test case
        new NumericCalc()
      case other =>
        throw new Exception(s"test case ${other} is not supported")
    }

    // defind streaming context
    val conf = new SparkConf().setMaster(config.master).setAppName(config.benchName)
    val ssc = new StreamingContext(conf, Milliseconds(config.batchInterval))
    ssc.checkpoint(config.checkpointPath)

    // add listener to collect static information.
    // val listener = new LatencyListener(ssc, config)
    // ssc.addStreamingListener(listener)

    val lines: DStream[(String, String)] = if (config.directMode) {
      // direct mode with low level Kafka API
      KafkaUtils.createDirectStream[String, String, StringDecoder, StringDecoder](
        ssc, config.kafkaParams, Set(config.sourceTopic))

    } else {
      // receiver mode with high level Kafka API
      val kafkaInputs = (1 to config.receiverNumber).map{ _ =>
        KafkaUtils.createStream[String, String, StringDecoder, StringDecoder](
          ssc, config.kafkaParams, Map(config.sourceTopic -> config.threadsPerReceiver), config.storageLevel)
      }
      ssc.union(kafkaInputs)
    }

    // convent key from String to Long, it stands for event creation time.
    val parsedLines = lines.map{ case (k, v) => (k.toLong, v) }
    testCase.process(parsedLines, config)

    ssc.start()
    ssc.awaitTermination()
  }
}

File: src/streambench/stormbench/src/main/java/com/intel/hibench/streambench/storm/RunBench.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.hibench.streambench.storm;

import com.intel.hibench.streambench.common.ConfigLoader;
import com.intel.hibench.streambench.common.Platform;
import com.intel.hibench.streambench.common.metrics.KafkaReporter;
import com.intel.hibench.streambench.common.metrics.MetricsUtil;
import com.intel.hibench.streambench.storm.micro.*;
import com.intel.hibench.streambench.storm.micro.NumericCalc;
import com.intel.hibench.streambench.storm.trident.*;
import com.intel.hibench.streambench.storm.util.BenchLogUtil;
import com.intel.hibench.streambench.storm.util.StormBenchConfig;

public class RunBench {

  public static void main(String[] args) throws Exception {
    runAll(args);
  }

  public static void runAll(String[] args) throws Exception {

    if (args.length < 2)
      BenchLogUtil.handleError("Usage: RunBench <ConfigFile> <FrameworkName>");

    StormBenchConfig conf = new StormBenchConfig();

    ConfigLoader cl = new ConfigLoader(args[0]);
    boolean TridentFramework = false;
    if (args[1].equals("trident")) TridentFramework = true;

    conf.nimbus = cl.getProperty("hibench.streambench.storm.nimbus");
    conf.nimbusAPIPort = Integer.parseInt(cl.getProperty("hibench.streambench.storm.nimbusAPIPort"));
    conf.zkHost = cl.getProperty("hibench.streambench.zookeeper.host");
    conf.workerCount = Integer.parseInt(cl.getProperty("hibench.streambench.storm.worker_count"));
    conf.spoutThreads = Integer.parseInt(cl.getProperty("hibench.streambench.storm.spout_threads"));
    conf.boltThreads = Integer.parseInt(cl.getProperty("hibench.streambench.storm.bolt_threads"));
    conf.benchName = cl.getProperty("hibench.streambench.benchname");
    conf.recordCount = Long.parseLong(cl.getProperty("hibench.streambench.record_count"));
    conf.topic = cl.getProperty("hibench.streambench.topic_name");
    conf.consumerGroup = cl.getProperty("hibench.streambench.consumer_group");
    conf.readFromStart = Boolean.parseBoolean(cl.getProperty("hibench.streambench.storm.read_from_start"));
    conf.ackon = Boolean.parseBoolean(cl.getProperty("hibench.streambench.storm.ackon"));
    conf.nimbusContactInterval = Integer.parseInt(cl.getProperty("hibench.streambench.storm.nimbusContactInterval"));

    String brokerList = cl.getProperty("hibench.streambench.brokerList");
    long recordPerInterval = Long.parseLong(cl.getProperty("hibench.streambench.prepare.periodic.recordPerInterval"));
    int intervalSpan = Integer.parseInt(cl.getProperty("hibench.streambench.prepare.periodic.intervalSpan"));
    String reporterTopic = MetricsUtil.getTopic(Platform.STORM, conf.topic, recordPerInterval, intervalSpan);
    conf.latencyReporter = new KafkaReporter(reporterTopic, brokerList);

    String benchName = conf.benchName;

    BenchLogUtil.logMsg("Benchmark starts... " + "  " + benchName +
            "   Frameworks:" + (TridentFramework ? "Trident" : "Storm"));

    if (TridentFramework) { // For trident workloads
      if (benchName.equals("wordcount")) {
        conf.separator = cl.getProperty("hibench.streambench.separator");
        TridentWordcount wordcount = new TridentWordcount(conf);
        wordcount.run();
      } else if (benchName.equals("identity")) {
        TridentIdentity identity = new TridentIdentity(conf);
        identity.run();
      } else if (benchName.equals("sample")) {
        conf.prob = Double.parseDouble(cl.getProperty("hibench.streambench.prob"));
        TridentSample sample = new TridentSample(conf);
        sample.run();
      } else if (benchName.equals("project")) {
        conf.separator = cl.getProperty("hibench.streambench.separator");
        conf.fieldIndex = Integer.parseInt(cl.getProperty("hibench.streambench.field_index"));
        TridentProject project = new TridentProject(conf);
        project.run();
      } else if (benchName.equals("grep")) {
        conf.pattern = cl.getProperty("hibench.streambench.pattern");
        TridentGrep grep = new TridentGrep(conf);
        grep.run();
      } else if (benchName.equals("distinctcount")) {
        conf.separator = cl.getProperty("hibench.streambench.separator");
        conf.fieldIndex = Integer.parseInt(cl.getProperty("hibench.streambench.field_index"));
        TridentDistinctCount distinct = new TridentDistinctCount(conf);
        distinct.run();
      } else if (benchName.equals("statistics")) {
        conf.separator = cl.getProperty("hibench.streambench.separator");
        conf.fieldIndex = Integer.parseInt(cl.getProperty("hibench.streambench.field_index"));
        TridentNumericCalc numeric = new TridentNumericCalc(conf);
        numeric.run();
      }
    } else { // For storm workloads
      if (benchName.equals("identity")) {
        Identity identity = new Identity(conf);
        identity.run();
      } else if (benchName.equals("project")) {
        conf.separator = cl.getProperty("hibench.streambench.separator");
        conf.fieldIndex = Integer.parseInt(cl.getProperty("hibench.streambench.field_index"));
        ProjectStream project = new ProjectStream(conf);
        project.run();
      } else if (benchName.equals("sample")) {
        conf.prob = Double.parseDouble(cl.getProperty("hibench.streambench.prob"));
        SampleStream sample = new SampleStream(conf);
        sample.run();
      } else if (benchName.equals("wordcount")) {
        conf.separator = cl.getProperty("hibench.streambench.separator");
        Wordcount wordcount = new Wordcount(conf);
        wordcount.run();
      } else if (benchName.equals("grep")) {
        conf.pattern = cl.getProperty("hibench.streambench.pattern");
        GrepStream grep = new GrepStream(conf);
        grep.run();
      } else if (benchName.equals("statistics")) {
        conf.separator = cl.getProperty("hibench.streambench.separator");
        conf.fieldIndex = Integer.parseInt(cl.getProperty("hibench.streambench.field_index"));
        NumericCalc numeric = new NumericCalc(conf);
        numeric.run();
      } else if (benchName.equals("distinctcount")) {
        conf.separator = cl.getProperty("hibench.streambench.separator");
        conf.fieldIndex = Integer.parseInt(cl.getProperty("hibench.streambench.field_index"));
        DistinctCount distinct = new DistinctCount(conf);
        distinct.run();
      } else if (benchName.equals("statistics")) {
        conf.separator = cl.getProperty("hibench.streambench.separator");
        conf.fieldIndex = Integer.parseInt(cl.getProperty("hibench.streambench.field_index"));
        NumericCalcSep numeric = new NumericCalcSep(conf);
        numeric.run();
      }
    }
  }
}
